R/expr-meta.R

```diff
@@ -67,16 +103,180 @@ expr_meta__as_selector <- function() {
+#' in one Polars version may not be deserializable in another Polars version.
+#'
+#' @inherit as_polars_expr return
+#' @examplesIf requireNamespace("jsonlite", quietly = TRUE)
+#' # Serialize the expression into a binary representation.
+#' expr <- pl$col("foo")$sum()$over("bar")
+#' bytes <- expr$meta$serialize()
+#' rawToChar(bytes)
+#'
+#' pl$deserialize_expr(bytes)
+#'
+#' # Serialize into json
+#' expr$meta$serialize(format = "json") |>
+#'   jsonlite::prettify()
+expr_meta_serialize <- function(file = NULL, ..., format = c("binary", "json")) {
   wrap({
     check_dots_empty0(...)
 
     format <- arg_match0(format, c("binary", "json"))
 
-    switch(format,
+    serialized <- switch(format,
       binary = self$`_rexpr`$serialize_binary(),
       json = self$`_rexpr`$serialize_json(),
       abort("Unreachable")
     )
+
+    if (is.null(file)) {
+      return(serialized)
+    }
+
+    if (format == "json") {
+      writeLines(serialized, con = file)
+    } else {
+      writeBin(serialized, con = file)
+    }
+  })
+}
+
+#' Indicate if this expression is the same as another expression
+#'
+#' @inherit as_polars_expr return
+#' @examples
+#' foo_bar <- pl$col("foo")$alias("bar")
+#' foo <- pl$col("foo")
+#' foo_bar$meta$eq(foo)
+#'
+#' foo_bar2 <- pl$col("foo")$alias("bar")
+#' foo_bar$meta$eq(foo_bar2)
+expr_meta_eq <- function(other) {
+  self$`_rexpr`$meta_eq(as_polars_expr(other)$`_rexpr`) |>
+    wrap()
+}
+
+#' Indicate if this expression is not the same as another expression
+#'
+#' @inherit as_polars_expr return
+#' @examples
+#' foo_bar <- pl$col("foo")$alias("bar")
+#' foo <- pl$col("foo")
+#' foo_bar$meta$ne(foo)
+#'
+#' foo_bar2 <- pl$col("foo")$alias("bar")
+#' foo_bar$meta$ne(foo_bar2)
+expr_meta_ne <- function(other) {
+  !self$eq(other)
+}
+
+#' Get a list with the root column name
+#'
+#' @inherit as_polars_expr return
+#' @examples
+#' e <- pl$col("foo") * pl$col("bar")
+#' e$meta$root_names()
+#'
+#' e_filter <- pl$col("foo")$filter(pl$col("bar") == 13)
+#' e_filter$meta$root_names()
+#'
+#' e_sum_over <- pl$sum("foo")$over("groups")
+#' e_sum_over$meta$root_names()
+#'
+#' e_sum_slice <- pl$sum("foo")$slice(pl$len() - 10, pl$col("bar"))
+#' e_sum_slice$meta$root_names()
+expr_meta_root_names <- function() {
+  self$`_rexpr`$meta_root_names()
+}
+
+#' Format the expression as a tree
+#'
+#' @inheritParams rlang::check_dots_empty0
+#' @param return_as_string Return the tree as a character vector? If `FALSE`
+#' (default), the tree is printed in the console.
+#'
+#' @inherit as_polars_expr return
+#' @examples
+#' my_expr <- (pl$col("foo") * pl$col("bar"))$sum()$over(pl$col("ham")) / 2
+#' my_expr$meta$tree_format()
+expr_meta_tree_format <- function(..., return_as_string = FALSE) {
+  wrap({
+    check_dots_empty0(...)
+    out <- self$`_rexpr`$meta_tree_format()
+    if (isTRUE(return_as_string)) {
+      out
+    } else {
+      cat(out)
+    }
```

Comment:
I'm not convinced about that, that's precisely what this argument is about. It would be counter-intuitive to have `return_as_string = FALSE` but still return a string (even if it's not printed in the console).

Reply:
Hmmm, I don't think the API of this function is so beautiful. Why do we need to specify a long argument like `return_as_string=TRUE` to get the output?
If the function returns a string from the beginning, we can just connect it to `cat()`.

There is a similar function `<LazyFrame>$explain()`, which always returns a string.
https://github.com/eitsupi/neo-r-polars/blob/c7461be38ac208d1275c8bdf78a2c2be6d896367/R/lazyframe-frame.R#L188

Reply:
I don't mind removing all the args and just return a string, I'll change that